tests/ignite/metrics/test_confusion_matrix.py

```diff
@@ -55,145 +55,97 @@ def test_multiclass_wrong_inputs():
+        else:
+            n_iters = y.shape[0] // batch_size + 1
+            for i in range(n_iters):
+                idx = i * batch_size
+                cm.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))
 
-        # 2-classes
-        num_classes = 2
-        cm = ConfusionMatrix(num_classes=num_classes)
-        y_pred = torch.rand(4, num_classes)
-        y = torch.randint(0, num_classes, size=(4,)).long()
-        cm.update((y_pred, y))
         np_y_pred = y_pred.numpy().argmax(axis=1).ravel()
         np_y = y.numpy().ravel()
         assert np.all(confusion_matrix(np_y, np_y_pred, labels=list(range(num_classes))) == cm.compute().numpy())
 
-        # Batched Updates
-        num_classes = 5
-        cm = ConfusionMatrix(num_classes=num_classes)
-
-        y_pred = torch.rand(100, num_classes)
-        y = torch.randint(0, num_classes, size=(100,)).long()
-
-        batch_size = 16
-        n_iters = y.shape[0] // batch_size + 1
-
-        for i in range(n_iters):
-            idx = i * batch_size
-            cm.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))
-
-        np_y = y.numpy().ravel()
-        np_y_pred = y_pred.numpy().argmax(axis=1).ravel()
-        assert np.all(confusion_matrix(np_y, np_y_pred, labels=list(range(num_classes))) == cm.compute().numpy())
-
     # check multiple random inputs as random exact occurencies are rare
     for _ in range(10):
-        _test_N()
+
+        def get_test_cases():
+            return [
+                (torch.rand(10, 4), torch.randint(0, 4, size=(10,)).long(), 4, 1),
+                (torch.rand(4, 10), torch.randint(0, 10, size=(4,)).long(), 10, 1),
+                (torch.rand(4, 2), torch.randint(0, 2, size=(4,)).long(), 2, 1),
+                # batched updates
+                (torch.rand(100, 5), torch.randint(0, 5, size=(100,)).long(), 5, 16),
+            ]
+
+        # check multiple random inputs as random exact occurencies are rare
+        for _ in range(10):
+            for y_pred, y, num_classes, batch_size in get_test_cases():
+                _test_N(y_pred, y, num_classes, batch_size)
 
 
 def test_multiclass_input_NL():
     # Multiclass input data of shape (N, L)
-    def _test_NL():
-        num_classes = 4
-        cm = ConfusionMatrix(num_classes=num_classes)
-
-        y_pred = torch.rand(10, num_classes, 5)
-        y = torch.randint(0, num_classes, size=(10, 5)).long()
-        cm.update((y_pred, y))
-        np_y_pred = y_pred.numpy().argmax(axis=1).ravel()
-        np_y = y.numpy().ravel()
-        assert np.all(confusion_matrix(np_y, np_y_pred, labels=list(range(num_classes))) == cm.compute().numpy())
+    def _test_NL(y_pred, y, num_classes, batch_size):
 
-        num_classes = 10
         cm = ConfusionMatrix(num_classes=num_classes)
-        y_pred = torch.rand(4, num_classes, 5)
-        y = torch.randint(0, num_classes, size=(4, 5)).long()
-        cm.update((y_pred, y))
+        if batch_size == 1:
```

Comment:
Same for this check

Reply:
@Moh-Yakoub let's code this test as here : https://github.com/pytorch/ignite/blob/a82327e1d908f122c359b9717a8003a45cf850e4/tests/ignite/metrics/test_accuracy.py#L64

```python

    def _test(y_pred, y, cm, batch_size):
         cm.reset()
         cm.update((y_pred, y))
         ...
         if batch_size > 1:
             ...

    for _ in range(5):
        for y_pred, y, num_classes, batch_size in get_test_cases():
            cm = ConfusionMatrix(num_classes=num_classes)
            _test(y_pred, y, cm, batch_size)

```

Reply:
@Moh-Yakoub please update it

Reply:
@vfdev-5 I am not sure your code does the same check. The original test case looks like this
``
        cm = ConfusionMatrix(num_classes=num_classes)

        y_pred = torch.rand(100, num_classes)
        y = torch.randint(0, num_classes, size=(100,)).long()

        batch_size = 16
        n_iters = y.shape[0] // batch_size + 1

        for i in range(n_iters):
            idx = i * batch_size
            cm.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))``

If I try to do as you suggested, 
``
      cm.update((y_pred, y))
        if batch_size > 1:
            n_iters = y.shape[0] // batch_size + 1
            for i in range(n_iters):
                idx = i * batch_size
                cm.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))``


That means I am making a different sequence of update to the `cm` metric. My tests are failing because of that as well. not sure if i am missing anything, but how is this supposed to produce the same results if we have multiple update sequences.

Reply:
@Moh-Yakoub yes, you are right, the first update should be under the condition, it does not make much sense to put the same sequence twice when `batch_size > 1`. Thanks a lot for pointing out that !
How about that:
```python
    def _test(y_pred, y, cm, batch_size):
         cm.reset()
         ...
         if batch_size > 1:
             ... batched updates
         else:   
             cm.update((y_pred, y))   # single update
         ...

    for _ in range(5):
        for y_pred, y, num_classes, batch_size in get_test_cases():
            cm = ConfusionMatrix(num_classes=num_classes)
            _test(y_pred, y, cm, batch_size)
```
What do you think ?

@KickItLikeShika we have to take that into account for your PR on Accuracy, Recall, Precision etc... and fix that once we are agreed on the way to do that.

Reply:
@vfdev-5 that's a good idea! Thank you @Moh-Yakoub for such great ideas!